core/trino-main/src/main/java/io/trino/operator/DeleteAndInsertMergeProcessor.java

```diff
@@ -0,0 +1,289 @@
+        pageTypesBuilder.addAll(dataColumnTypes);
+        pageTypesBuilder.add(INTEGER);
+        List<Type> pageTypes = pageTypesBuilder.build();
+        PageBuilder pageBuilder = new PageBuilder(totalPositions, pageTypes);
+        int resultPosition = 0;
+        int rowIdPosition = 0;
+        int[] rowIdPositions = new int[totalPositions];
+        for (int position = 0; position < originalPositionCount; position++) {
+            long operation = INTEGER.getLong(operationChannelBlock, position);
+            if (operation != DEFAULT_CASE_OPERATION_NUMBER) {
+                if (operation == DELETE_OPERATION_NUMBER || operation == UPDATE_OPERATION_NUMBER) {
+                    addDeleteRow(pageBuilder, inputPage, position);
+                    rowIdPositions[resultPosition] = rowIdPosition;
+                    resultPosition++;
+                }
+                if (operation == INSERT_OPERATION_NUMBER || operation == UPDATE_OPERATION_NUMBER) {
+                    addInsertRow(pageBuilder, mergeCaseBlocks, position);
+                    rowIdPositions[resultPosition] = 0;
+                    resultPosition++;
+                }
+            }
+            if (INTEGER.getLong(rowMatchedBlock, position) > 0) {
+                rowIdPosition++;
+            }
+        }
+        checkArgument(resultPosition == totalPositions, "resultPosition (%s) is not equal to totalPositions (%s)", resultPosition, totalPositions);
+        Page pageWithoutRowIdBlock = pageBuilder.build();
+        int pageChannels = pageWithoutRowIdBlock.getChannelCount();
+        Block[] blocksWithRowId = new Block[pageChannels + 1];
+        for (int channel = 0; channel < pageChannels; channel++) {
+            blocksWithRowId[channel] = pageWithoutRowIdBlock.getBlock(channel);
+        }
+        blocksWithRowId[pageChannels] = makeOutputRowIdBlock(rowIdBlock, totalPositions, rowIdPositions);
+
+        return new Page(blocksWithRowId);
+    }
+
+    private static Block makeOutputRowIdBlock(Block rowIdBlock, int totalPositions, int[] rowIdPositions)
+    {
+        Block underlyingBlock = getUnderlyingBlock(rowIdBlock);
+        if (rowIdBlock.allPositionsAreNull()) {
+            return getAllNullsRowIdBlock(rowIdBlock, underlyingBlock, totalPositions);
+        }
+        else {
+            if (underlyingBlock instanceof RowBlock) {
+                List<Block> newRowIdChildrenBuilder = new ArrayList<>();
+                rowIdBlock.getChildren().stream()
+                        .map(block -> block.getPositions(rowIdPositions, 0, totalPositions))
+                        .forEach(newRowIdChildrenBuilder::add);
+                return RowBlock.fromFieldBlocks(
+                        totalPositions,
+                        Optional.empty(),
+                        newRowIdChildrenBuilder.toArray(new Block[] {}));
+            }
+            else {
+                return rowIdBlock.getPositions(rowIdPositions, 0, totalPositions);
```

Comment:
Why `RowBlock` is special-cased here?
What if `underlyingBlock` is a DictionaryBlock over a RowBlock? Would it require special-casing as well?

Reply:
I had endless trouble with this, and it's one of the main things I hoped review would shed light on.

I had hoped that I could just call `rowIdBlock.getPositions(...)` and end up with a consistent view of the resulting block.  However, when I tried that, way downstream in the `Driver` I would see out-of-range array references.  My assumption is that I'm doing something wrong, but I wasn't successful debugging the problem.

Reply:
> I had endless trouble with this, and it's one of the main things I hoped review would shed light on.

Sorry that i cannot help. Add a TODO comment here, warning the reader we don't exactly know why it's written the way it's written